[server] Authentication nonce tweak

e2r3p13 opened the pull request:
Fix a potential authentication bypass:

During the authentication phase, we send an encrypted nonce, with a prepended IV, needed by the client to decrypt the nonce with the shared key. The client then have to reencrypt the nonce with another IV and send it back to the server.
The server then compares the received nonce with the sent one and if they match, the client is authenticated.

The issue here is that an attacker can connect, listen for the encrypted nonce from the server, and send back the exact same message. This message contains the right nonce, encrypted with the shared key and a valid IV.

Tweaking the nonce in a deterministic way before sending it back ensures that someone who doesn't know the shared key can't successfully bypass the authentication process.

The way we tweak the nonce is the following:

- We split the nonce in two halfs (8 bytes long each)
- We invert those halfs

In python, it would give: tweaked_nonce = nonce[8:] + nonce[:8]

Signed-off-by: Loic Buckwell <lfalkau@student.42.fr>

e2r3p13 committed:
[server] Authentication nonce tweak

Fix a potential authentication bypass:

During the authentication phase, we send an encrypted nonce,
with a prepended IV, needed by the client to decrypt the nonce
with the shared key. The client then have to reencrypt the nonce
with another IV and send it back to the server.
The server then compares the received nonce with the sent one
and if they match, the client is authenticated.

The issue here is that an attacker can connect, listen for
the encrypted nonce from the server, and send back the exact same
message. This message contains the right nonce, encrypted with the
shared key and a valid IV.

Tweaking the nonce in a deterministic way before sending it back ensures
that someone who doesn't know the shared key can't successfully bypass the
authentication process.

The way we tweak the nonce is the following:

- We split the nonce in two halfs (8 bytes long each)
- We invert those halfs

In python, it would give: tweaked_nonce = nonce[8:] + nonce[:8]

Signed-off-by: Loic Buckwell <lfalkau@student.42.fr>

e2r3p13 commented:
Maybe I should have mentioned that in the commit message, but the reason I prefer this tweak is that it is easy to implement with most languages, and every bit is moved.

kerneis-anssi commented on server/protocol.go:
```diff
@@ -144,7 +144,12 @@ func authentication(session *Session) error {
+		close(session.ch)
+		return errors.New("Authentication failure: Invalid nonce")
+	}
+	tweaked_nonce := append(rcvd_nonce.Bytes[8:], rcvd_nonce.Bytes[:8]...)
```
Please add a comment explaining the rationale.

kerneis-anssi commented on server/protocol.go:
```diff
@@ -144,7 +144,12 @@ func authentication(session *Session) error {
+		return errors.New("Authentication failure: Invalid nonce")
+	}
+	tweaked_nonce := append(rcvd_nonce.Bytes[8:], rcvd_nonce.Bytes[:8]...)
+	if bytes.Equal(nonce.Bytes, tweaked_nonce) == false {
```
Can we have a test for this in protocol_test.go?

e2r3p13 committed:
[tests] Add authentication tests

Add three tests for the authentication protocol step.
Also replace Fatal* calls by Error* ones in test goroutines,
to fix copmiler warnings.

Signed-off-by: Loic Buckwell <lfalkau@student.42.fr>

e2r3p13 commented:
The CI fails because when testing the authentication function, the tests must be run as root. This is because the authentication function uses the TPM to generate random.

kerneis-anssi commented:
> The CI fails because when testing the authentication function, the tests must be run as root. This is because the authentication function uses the TPM to generate random.

I think tests are run as root, but I guess there's no TPM on the CI…

e2r3p13 commented:
Another solution to avoid the need of the nonce tweak would be to ensure on both sides that the IV was generated on the other side. This could be done by setting the first IV bit to 0 for a server message and to 1 for a client message. Both parts would then check that bit before processing the message.

Finally, a third solution would be to completely avoid this nonce, and replace it with the client sending a predictable but always different data, e.g. a timestamp, encrypted with the shared key. As the encryption is authenticated, if the server is able to decrypt it and to match the timestamp, it would be enough to authenticate a genuine client.

Let me know what you think about it.
